keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -656,73 +661,287 @@ ways.  One is that we are dispatching requests to be served rather
+original fantasy, this virtual world uses the same clock as the real
+world.  Whenever a packet finishes being sent in the real world, the
+next packet to be transmitted is the unsent one that will finish being
+sent soonest in the virtual world.  If there is a tie among several,
+we pick the one whose queue is next in round-robin order (following
+the queue last picked).
+
+We can define beginning and end times (B and E) for transmission of
+the j'th packet of queue i in the virtual world, with the following
+equations.
+
+```
+B(i,j) = max(E(i,j-1), t_arrive(i,j))
+Integral[from tau=B(i,j) to tau=E(i,j)] mu(i,tau) dtau = len(i,j)
+```
+
+This has a practical advantage over the original story: the integrals
+are only over the lifetime of a single request's service --- rather
+than over the lifetime of the server.  This makes it easier to use
+floating point representations with sufficient precision.
+
+Note that computing an E value before it has arrived requires
+predicting the course of `mu(i,t)` from now until E arrives.  However,
+because all the non-empty queues have the same value for `mu(i,t)`
+(i.e., `mu_fair(t)`), we can safely make whatever assumption we want
+without distorting the dispatching choice --- all non-empty queues are
+affected equally, so even wildly wrong guesses don't change the ordering.
+
+The correspondence with the original telling of the fair queuing story
+goes as follows.  Equate
+
+```
+S(i,j) = R(B(i,j))
+F(i,j) = R(E(i,j))
+```
+
+The recurrence relations for S and F correspond to the recurrence
+relations for B and E.
+
+```
+S(i,j) = R(B(i,j))
+       = R(max(E(i,j-1), t_arrive(i,j)))
+       = max(R(E(i,j-1)), R(t_arrive(i,j)))
+       = max(F(i,j-1), R(t_arrive(i,j)))
+```
+
+(R and max commute because both are monotonically non-decreasing).
+
+Note that `mu_fair(t)` is exactly the same as `dR/dt` in the original
+story.  So we can reason as follows.
+
+```
+Integral[tau=B(i,j) to tau=E(i,j)] mu(i,tau) dtau = len(i,j)
+
+Integral[tau=start to tau=E(i,j)] (dR/dt)(tau) dtau -
+Integral[tau=start to tau=B(i,j)] (dR/dt)(tau) dtau
+= len(i,j)
+
+R(E(i,j)) - R(B(i,j)) = len(i,j)
+
+F(i,j) - S(i,j) = len(i,j)
+
+F(i,j) = S(i,j) + len(i,j)
+```
+
+It is not necessary to track the B and E values for every unsent
+packet in the virtual world.  It suffices to track, for each queue
+`i`, the following things in the virtual world:
+- the contents of the queue
+- B of the packet currently being sent
+- bits of that packet sent so far: `Integral[from tau=B to tau=now] mu(i,tau) dtau`
+
+At any point in time the E of the packet being sent can be calculated,
+under the assumption that `mu(i,t)` will henceforth be constant, by
+adding to the current time the quotient of (remaining bits to send) /
+`mu(i,t)`.  The E of the packet after that can be calculated by
+further adding the quotient (size of packet) / `mu(i,t)`.  This E is
+the value used to pick the next packet to send in the real world.
```

Comment:
Oh wait, that's not necessarily right.  At dispatch time we care about the next packet that has not yet begin being transmitted in the real world.  That could be the one that is currently being sent in the virtual world.  I think it has to be one of these two, so we still have O(1) cost to compute the E of the correct packet.

Reply:
Revised with an eye towards correctness.